internal/configs/resource.go

```diff
@@ -813,56 +929,24 @@ func (m *ProviderConfigRefMapping) decodeStaticAlias(eval *StaticEvaluator, forE
 		})
 	}
 
-	// TODO/Oleksandr: Is static identifier valid if we evaluate for_each for alias?
-	instances, instancesDiags := eval.Evaluate(forEach, aliasStaticID)
+	instances, instancesDiags := eval.Evaluate(instanceExpr, aliasStaticID)
 	diags = diags.Extend(instancesDiags)
 	if instancesDiags.HasErrors() {
 		return diags
 	}
 
-	if !instances.CanIterateElements() {
-		panic("decodeStaticAlias must be called with for each expression: wrong expression?")
-	}
-
-	aliases := make(map[addrs.InstanceKey]string)
+	var aliases map[addrs.InstanceKey]string
+	var aliasDiags hcl.Diagnostics
 
-	instanceIter := instances.ElementIterator()
-	for instanceIter.Next() {
-		k, v := instanceIter.Element()
-
-		instanceEvalCtx := aliasEvalCtx.NewChild()
-		instanceEvalCtx.Variables = map[string]cty.Value{
-			"each": cty.ObjectVal(map[string]cty.Value{
-				"key":   k,
-				"value": v,
-			}),
-		}
-
-		aliasVal, aliasDiags := m.Alias.Value(instanceEvalCtx)
-		diags = diags.Extend(aliasDiags)
-		if aliasDiags.HasErrors() {
-			continue
-		}
-
-		if !aliasVal.Type().Equals(cty.String) {
-			// TODO/Oleksandr: add friendly error
-			return diags.Append(&hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Invalid alias",
-				Detail:   "",
-				Subject:  m.Alias.Range().Ptr(),
-			})
-		}
-
-		instanceKey, err := addrs.ParseInstanceKey(k)
-		if err != nil {
-			panic("decodeStaticAlias failed to parse for each instance keys: wrong expression?")
-		}
-
-		aliases[instanceKey] = aliasVal.AsString()
+	switch {
+	case hasEachRefInAlias:
+		aliases, aliasDiags = generateForEachAliases(instances, aliasEvalCtx, m.Alias)
+	case hasCountRefInAlias:
+		aliases, aliasDiags = generateCountAliases(instances, aliasEvalCtx, m.Alias)
 	}
 
-	if !diags.HasErrors() {
+	diags = diags.Extend(aliasDiags)
+	if !aliasDiags.HasErrors() {
```

Comment:
this is more of a tidy up on the existing code. should we tweak this last bit to be more idiomatic? 

```
diags = diags.Extend(aliasDiags)
if aliasDiags.HasErrors() {
    return diags
}

m.Aliases = aliases
return diags
```

Reply:
Thanks! Made it more idiomatic in [8075c3f](https://github.com/opentofu/opentofu/pull/1959/commits/8075c3f26301c29062c5e2568f6bba9a762b71c5)